tools/system_libs.py

```diff
@@ -286,325 +605,391 @@ def create_libc(libname):
+  asmjs_pthreads_symbols = read_symbols(shared.path_from_root('system', 'lib', 'asmjs_pthreads.symbols'))
+  stub_pthreads_symbols = read_symbols(shared.path_from_root('system', 'lib', 'stub_pthreads.symbols'))
+
+  def get_symbols(self):
+    symbols = self.pthreads_symbols if self.is_mt else self.stub_pthreads_symbols
+    if self.is_mt and not shared.Settings.WASM_BACKEND:
+      symbols += self.asmjs_pthreads_symbols
+    return symbols
+
+
+class CompilerRTWasmLibrary(NoBCLibrary):
+  cflags = ['-O2', '-fno-builtin']
+  force_object_files = True
+
+  def can_build(self):
+    return shared.Settings.WASM_BACKEND
+
+
+class libcompiler_rt_wasm(CompilerRTWasmLibrary):
+  name = 'libcompiler_rt_wasm'
+
+  src_dir = ['system', 'lib', 'compiler-rt', 'lib', 'builtins']
+  src_files = ['addtf3.c', 'ashlti3.c', 'ashrti3.c', 'atomic.c', 'comparetf2.c',
+               'divtf3.c', 'divti3.c', 'udivmodti4.c',
+               'extenddftf2.c', 'extendsftf2.c',
+               'fixdfti.c', 'fixsfti.c', 'fixtfdi.c', 'fixtfsi.c', 'fixtfti.c',
+               'fixunsdfti.c', 'fixunssfti.c', 'fixunstfdi.c', 'fixunstfsi.c', 'fixunstfti.c',
+               'floatditf.c', 'floatsitf.c', 'floattidf.c', 'floattisf.c',
+               'floatunditf.c', 'floatunsitf.c', 'floatuntidf.c', 'floatuntisf.c', 'lshrti3.c',
+               'modti3.c', 'multc3.c', 'multf3.c', 'multi3.c', 'subtf3.c', 'udivti3.c', 'umodti3.c', 'ashrdi3.c',
+               'ashldi3.c', 'fixdfdi.c', 'floatdidf.c', 'lshrdi3.c', 'moddi3.c',
+               'trunctfdf2.c', 'trunctfsf2.c', 'umoddi3.c', 'fixunsdfdi.c', 'muldi3.c',
+               'divdi3.c', 'divmoddi4.c', 'udivdi3.c', 'udivmoddi4.c']
+
+  def get_files(self):
+    return super(libcompiler_rt_wasm, self).get_files() + [shared.path_from_root('system', 'lib', 'compiler-rt', 'extras.c')]
+
+
+class libc_rt_wasm(CompilerRTWasmLibrary, MuslInternalLibrary):
+  name = 'libc_rt_wasm'
+
+  def get_files(self):
+    return get_wasm_libc_rt_files()
+
+
+class libubsan_minimal_rt_wasm(CompilerRTWasmLibrary, MTLibrary):
+  name = 'libubsan_minimal_rt_wasm'
+
+  src_dir = ['system', 'lib', 'compiler-rt', 'lib', 'ubsan_minimal']
+  src_files = ['ubsan_minimal_handlers.cpp']
+
+
+class libsanitizer_common_rt_wasm(CompilerRTWasmLibrary, MTLibrary):
+  name = 'libsanitizer_common_rt_wasm'
+  depends = ['libc++abi']
+  js_depends = ['memalign']
+
+  cflags = ['-std=c++11']
+  src_dir = ['system', 'lib', 'compiler-rt', 'lib', 'sanitizer_common']
+  src_glob = '*.cc'
+  src_glob_exclude = ['sanitizer_common_nolibc.cc']
+
+
+class libubsan_rt_wasm(CompilerRTWasmLibrary, MTLibrary):
+  name = 'libubsan_rt_wasm'
+  depends = ['libsanitizer_common_rt_wasm']
+
+  includes = [['system', 'lib', 'compiler-rt', 'lib']]
+  cflags = ['-std=c++11', '-DUBSAN_CAN_USE_CXXABI']
+  src_dir = ['system', 'lib', 'compiler-rt', 'lib', 'ubsan']
+  src_glob = '*.cc'
+
+
+def calculate(temp_files, in_temp, stdout_, stderr_, forced=[]):
```

Comment:
This may not be appropriate for this PR, but it would be great if this function had a more descriptive name.